karapace/schema_reader.py

```diff
@@ -185,8 +162,37 @@ def get_schema_id(self, new_schema: TypedSchema) -> int:
             self.global_schema_id += 1
             return self.global_schema_id
 
-    def get_schema_id_if_exists(self, *, subject: Subject, schema: TypedSchema) -> Optional[SchemaId]:
-        return self._hash_to_schema_id_on_subject.get(hash((subject, schema.__str__())), None)
+    def get_schema_id_if_exists(self, *, subject: Subject, schema: ValidatedTypedSchema) -> Optional[SchemaId]:
+        sid = self._hash_to_schema_id_on_subject.get(hash((subject, schema.__str__())), None)
+        if sid is not None:
+            return sid
+        # In case if the hashing does not return id check with parsed schema if
+        # equivalent schema can be found. In Avro e.g. "string" == {"type":"string"}.
+        # If found set the equivalent schema to cache.
+        maybe_id = None
+        for existing_id, existing_schema in self.schemas.items():
```

Comment:
This is missing locking. `get_schema_id_if_exists` can be called from API, but schema reader thread handling Kafka messages may also mutate `self.schemas` during iteration

Reply:
Fixed, added a function in `schema_registry` to handle locking when called from `schema_registry_api`.

Reply:
This is now protected by `asyncio.Lock` but I think schema registry API and schema reader's message handler are running in two different threads. Therefore protection should be using `threading.Lock`?

Reply:
Done.